INT-4131: Add delayExpression to AMQP Outbounds

garyrussell opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4131

Specify an expression on the outbound endpoints to set the `x-delay` header when
using the RabbitMQ Delayed Message Exchange plugin.

garyrussell committed:
INT-4131: Add delayExpression to AMQP Outbounds

JIRA: https://jira.spring.io/browse/INT-4131

Specify an expression on the outbound endpoints to set the `x-delay` header when
using the RabbitMQ Delayed Message Exchange plugin.

artembilan commented on spring-integration-amqp/src/main/java/org/springframework/integration/amqp/outbound/AbstractAmqpOutboundEndpoint.java:
```diff
@@ -188,6 +192,31 @@ public void setLazyConnect(boolean lazyConnect) {
+	 * Set the SpEL expression to calculate the {@code x-delay} header when using the
+	 * RabbitMQ delayed message exchange plugin. By default, the {@link AmqpHeaders#DELAY}
+	 * header (if present) is mapped; setting the expression here overrides that value.
+	 * @param delayExpression the expression.
```
`@since 5.0`.
And no just `setDelay(int delay)` ?

artembilan commented on spring-integration-amqp/src/main/java/org/springframework/integration/amqp/outbound/AbstractAmqpOutboundEndpoint.java:
```diff
@@ -284,6 +313,13 @@ protected final void doInit() {
 					"A 'confirmCorrelationExpression' is required when specifying a 'confirmNackChannel'");
 		}
+		if (this.delayExpression != null) {
+			this.delayGenerator = new ExpressionEvaluatingMessageProcessor<Integer>(this.delayExpression,
```
I don't see reason in this extra object, but let it be: it is existing logic in this component.

artembilan commented on src/reference/asciidoc/amqp.adoc:
```diff
@@ -1149,7 +1149,14 @@ To configure a default user id for outbound messages, configure it on a `RabbitT
+[[amqp-delay]]
+=== Delayed Message Exchange
+
+Spring AMQP supports the http://docs.spring.io/spring-amqp/docs/1.6.3.RELEASE/reference/html/_reference.html#delayed-message-exchange[RabbitMQ Delayed Message Exchange Plugin].
```
I find this link as much better: http://docs.spring.io/spring-amqp/reference/html/_reference.html#delayed-message-exchange.
At least it is version-free.

artembilan reviewed (changes requested):
That's all.
Nothing critical, but still requires some explanation at least.
Thanks

garyrussell committed:
Polishing

- PR Comments
- Add `setDelay`
- Port `FunctionExpression` from DSL
- Add `SupplierExpression`

artembilan commented on spring-integration-amqp/src/main/java/org/springframework/integration/amqp/outbound/AbstractAmqpOutboundEndpoint.java:
```diff
@@ -193,10 +194,22 @@ public void setLazyConnect(boolean lazyConnect) {
+	 * @since 5.0
+	 */
+	public void setDelay(int delay) {
+		this.delayExpression = new SupplierExpression<>(() -> delay);
```
Good idea, of course, and it really has permissions to live :smile: 
But I think an existing `ValueExpression<V>` would be better here.

And since the setter is here already, WDYT about one more `delay` XSD attribute ?

In addition need to consider to move all this "strange" `Expression` implementations to SF SpEL module.

garyrussell replied:
I don't think it's necessary to add it to the XML since a literal numeric looks the same; with expressions that evaluate to String, you need quotes:

     delay-expression="12"

Vs.

    foo-expression="'bar'"

garyrussell replied:
After we merge this, we can open a JIRA for SF pointing to this code and see what Juergen thinks.

garyrussell replied:
I forgot about `ValueExpression` but, yes, `SupplierExpression` might still be useful so we might as well keep it.

artembilan replied:
Good observation about `delay-expression="12"`, but that is only a tip of the iceberg.
The SpEL parser produces `SpelExpression` anyway. It isn't so heavy though with just `IntLiteral` in the `ast`, but still.
I don't say that we should go with additional `delay` attribute, but keep in mind a SpEL nature would be good idea anyway.

Merging...

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/expression/SupplierExpression.java:
```diff
@@ -0,0 +1,184 @@
+ * Any {@link #setValue} operations and {@link #getValueType} related operations
+ * throw {@link EvaluationException}.
+ *
+ * @param <S> The evaluation context root object type.
```
Wrong JavaDoc:
```
@param name not found

 * @param <S> The evaluation context root object type.
```

garyrussell committed:
Javadoc Fixes

garyrussell committed:
Use ValueExpression for delay

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/9673a02c7da2924559fec0d845a8951f75c84cf0
